src/holon/models/util.py

```diff
@@ -48,6 +49,27 @@ def duplicate_model(obj, attrs={}):
     obj.save()
     return obj
 
+def duplicate_model_nomutate(source: Model) -> Model:
+    # shallow copy
+    destination = copy.copy(source)
+    destination.pk = None
+    destination.id = None
+
+    # for copying polymorphic models with multiple levels of inheritance
+    # https://stackoverflow.com/a/74999379/19602496
+    for field in source._meta.get_fields(include_parents=True):
```

Comment:
I copied this code from above in this same file. I'm not sure what it does but it looks like a hot loop.

Can it be removed?

Reply:
One idea is to cache the fields per object using properties (
`def fields(self) -> list[str]:
   if self._fields is None:
      self._fields = self._meta.get_fields(include_parents=True)
   return self._fields`

I would also advise to just adapt the `duplicate_model` function instead of copying it.

Reply:
> One idea is to cache the fields per object using properties

I will profile this again. If this loop shows up in the profile I will offer a separate pull request.

> I would also advise to just adapt the duplicate_model function instead of copying it, since without the attrs parameter the two functions (should) do the same. 

`duplicate_model` mutates it's input rather than create a new object. I don't want to mutate the original object because that makes Scenario.clone() more complex.

Both functions are sufficiently different in this next revision. If we decide to make the rules engine in-memory we won't need the original `duplicate_model`.

> Although, the way it is implemented right now the duplicate_model_nomutate just copies an object and clears its relations, but does not save a duplicate. If this is the intended behaviour,

This is intended so that I can do a batch insert.

>  I would rename the function to something like copy_and_clear_model

The clearing of the internal pointer and PK seems an implementation detail which doesn't need to be in the function name. 

> put the shared code between this function and duplicate_model in a function as well.

Done, created a function `clear_base_pointer(obj: Model)` with the shared code.

Reply:
> I will profile this again. If this loop shows up in the profile I will offer a separate pull request.

If it helps reduce db queries for cloning, why not add it to this PR?

> The clearing of the internal pointer and primary key seems an implementation detail which doesn't need to be in the function name.

If this is what the function does, the name should reflect this in some sense in my opinion. If it does not duplicate, naming it as such might be confusing.

Reply:
> If it helps reduce db queries for cloning, why not add it to this PR?

Because I want this merged ASAP so we can unblock Jorrit

Reply:
Okay, I was not aware this was an urgent PR. I was just asked to do a code review so I reviewed the code ;-)
If this works and it unblocks Jorrit i'd say ready to merge, we can discuss the details later.

Reply:
Ok I'm definitely willing to work it out.